✨🏗 Upgrade closure compiler to v20190301

rsimha opened the pull request:
This PR upgrades the version of closure compiler used to minify the AMP runtime to v20190301.

**Highlights:**
- Updated `compiler.jar` with a custom built version based on [v20190301](https://github.com/google/closure-compiler/wiki/Releases#march-1-2019-v20190301)
    - Reverts a [breaking change](https://github.com/google/closure-compiler/commit/acea045bcc566c8d1ba74bc8f63fbb1568a18c2b) (required until https://github.com/google/closure-compiler/issues/3041 is fixed)
- Updated `compiler-and-tests.jar` with a custom built version based on [v20190301](https://github.com/google/closure-compiler/wiki/Releases#march-1-2019-v20190301)
    - Follows the steps outlined [here](https://github.com/ampproject/amphtml/pull/18552#issue-220501644)
- Updated documentation for custom build changes in [README.amp](https://github.com/ampproject/amphtml/blob/2731c69e684df7421009f4d956da38da436375a6/third_party/closure-compiler/README.amp)
- Updated `runner.jar` after fixing code and tests to be compatible with latest compiler code, and to correctly strip out asserts
- Numerous fixes for missing / incorrect `@type` annotations in AMP source code
- Fixes for incorrectly annotated default localization exports in `amp-story` and `amp-story-auto-ads`
- Assorted fixes to make single-pass and multi-pass builds and tests green

**For posterity, here is a list of things I've reviewed / verified / tested.**

**Code review:**
- Self-reviewed all 240+ files in this PR to make sure there are no stray changes
- Re-read all 200+ review comments to make sure they were satisfactorily addressed and resolved
- Filed issues for future clean up tasks / feature requests

**Local tests:**
- Ran `gulp check-types` to make sure there are no type errors
- Ran `gulp dist --fortesting` to make sure minified multi-pass build is healthy
    - Ran `gulp serve --compiled` and manually tested several components in multi-pass mode
- Ran `gulp dist --fortesting --single_pass --pseudo_names` to make sure minified single-pass build is healthy
    - Ran `gulp serve --compiled` and manually tested several components in single-pass mode
- Ran `gulp test --local-changes --headless` to make sure all unit tests affected by this PR are green

**Travis tests:**
- Checks (`presubmit`, `lint`, `dep-check`, and `check-types`) are green ([link](https://travis-ci.org/ampproject/amphtml/jobs/518418541#L640-L745))
- Un-minified build is green ([link](https://travis-ci.org/ampproject/amphtml/jobs/518418540#L640-L644))
- Visual tests are green with zero diffs due to this PR ([link](https://percy.io/ampproject/amphtml/builds/1719711))
- All unit tests are green ([link](https://travis-ci.org/ampproject/amphtml/jobs/518418545#L3772))
- All integration tests are green ([link](https://travis-ci.org/ampproject/amphtml/jobs/518418545#L3742))
- Minified single-pass build and tests are green ([link](https://travis-ci.org/ampproject/amphtml/jobs/518418543#L691))

**AMP Release schedule**
- AMP Conference 2019 is next week. We're doing a [release](https://github.com/ampproject/amphtml/issues/21788) on Tuesday, Apr 16 using this week's canary cut, which was based on the [1904091426070](https://github.com/ampproject/amphtml/releases/tag/1904091426070) tag.
- This PR was merged after the AMP Conf release tag, so the main AMP Conf release is unaffected by any of these changes.

Partial fix for #18748
Follow up to #18552
Follow up to #18609
Follow up to #18794
Follow up to #19449
Follow up to #20056

rsimha commented:
At long last, this PR is ready for preliminary review. We are now at a point where `gulp dist` passes in multi-pass and single-pass mode, and `gulp check-types`, `gulp lint`, and `gulp presubmit` are all green. I'm still verifying the changes I've made, and working through test failures.

Sending this out to all TLs and those familiar with closure compiler. Feel free to assign directories / files to others on your team. For easy reviewing, I've split off the changes into separate commits as far as possible.

cramforce commented:
Woohoo! Makes me super happy that we are so close to head!

rsimha commented:
**Good news:** All tests pass in multi-pass mode ([link](https://travis-ci.org/ampproject/amphtml/jobs/512851531))
**Bad news:** Several test failures in single-pass mode ([link](https://travis-ci.org/ampproject/amphtml/jobs/512851529#L963))

Paging @erwinmombay to see if I'm missing something obvious that's causing the single-pass failures

kristoferbaxter commented on extensions/amp-apester-media/0.1/amp-apester-media.js:
```diff
@@ -139,15 +139,15 @@ class AmpApesterMedia extends AMP.BaseElement {
+    this.mediaAttribute_ = /** @type {string} */ (userAssert(
         this.element.getAttribute('data-apester-media-id') ||
-        (this.random_ = this.element.getAttribute(
+        (this.random_ = this.element.hasAttribute(
```
Intentional change? From string to boolean?

rsimha replied:
Yes, there was a mismatch earlier, that I discovered via `gulp check-types`.

jridgewell replied:
The `playlist` property is only used as a boolean.

rsimha replied:
@aghassemi @jridgewell let me know what the desired fix is here

rsimha replied:
Done in https://github.com/ampproject/amphtml/commit/a0d8a000bb441d47af92effd8a069f16e3675d5f.

kristoferbaxter commented on extensions/amp-riddle-quiz/0.1/amp-riddle-quiz.js:
```diff
@@ -31,8 +31,8 @@ export class AmpRiddleQuiz extends AMP.BaseElement {
 
-    /** @private {?number} */
-    this.riddleId_ = null;
+    /** @private {string} */
```
Intentional? Will data['riddleId'] return a string or a number?

rsimha replied:
Yes, intentional. Without this change:
```js
extensions/amp-riddle-quiz/0.1/amp-riddle-quiz.js:72: ERROR - assignment to property riddleId_ of AmpRiddleQuiz
found   : string
required: (null|number)
    this.riddleId_ = userAssert(
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^

extensions/amp-riddle-quiz/0.1/amp-riddle-quiz.js:89: ERROR - actual parameter 1 of encodeURIComponent does not match formal parameter
found   : (null|number)
required: string
    iframe.src = 'https://www.riddle.com/a/iframe/' + encodeURIComponent(this.riddleId_);
                                                                         ^^^^^^^^^^^^^^
```

kristoferbaxter commented on extensions/amp-story-auto-ads/0.1/amp-story-auto-ads.js:
```diff
@@ -161,10 +161,10 @@ export class AmpStoryAutoAds extends AMP.BaseElement {
     super(element);
 
-    /** @private {?../../amp-story/0.1/amp-story.AmpStory} */
+    /** @private {?../../amp-story/1.0/amp-story.AmpStory} */
```
Remnant of moving from `0.1` to `1.0`? Is this change required as part of moving to a new Closure Compiler or just convenient to make in this PR?

rsimha replied:
Yes, looks like a remnant that wasn't taken care of earlier. Without this change:
```js
extensions/amp-story-auto-ads/0.1/amp-story-auto-ads.js:263: ERROR - assignment to property navigationState_ of AmpStoryAutoAds
found   : NavigationState$$module$extensions$amp_story$0_1$navigation_state
required: (NavigationState$$module$extensions$amp_story$1_0$navigation_state|null)
    this.navigationState_ = this.ampStory_.getNavigationState();
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

extensions/amp-story-auto-ads/0.1/amp-story-auto-ads.js:732: ERROR - actual parameter 1 of AmpStoryAutoAds$$module$extensions$amp_story_auto_ads$0_1$amp_story_auto_ads.prototype.nextPageNoAd_ does not match formal parameter
found   : AmpStoryPage$$module$extensions$amp_story$0_1$amp_story_page
required: (AmpStoryPage$$module$extensions$amp_story$1_0$amp_story_page|null)
        || this.nextPageNoAd_(pageBeforeAd)
                              ^^^^^^^^^^^^

extensions/amp-story-auto-ads/0.1/amp-story-auto-ads.js:750: ERROR - Property getPageIndexById never defined on AmpStory
    const pageNumber = this.ampStory_.getPageIndexById(pageBeforeAdId);
                                      ^^^^^^^^^^^^^^^^
```

kristoferbaxter commented on extensions/amp-story/1.0/_locales/af.js:
```diff
@@ -20,7 +20,7 @@ import {LocalizedStringBundleDef, LocalizedStringId} from '../localization';
  * @const {!LocalizedStringBundleDef}
  */
-export default /** @const {!LocalizedStringBundleDef} */ ({
+const strings = {
```
Are these changes required? Is there an impact on build size by creating a reference that's exported versus inlining?

rsimha replied:
Yes, they're required. The typedefs in the current version of the code were added during a previous upgrade of closure compiler https://github.com/ampproject/amphtml/pull/18552/commits/f949f8f36f045314b270ffd85ad0f8d5735a35a7. It turns out that this isn't a valid way to annotate default exports (https://github.com/jsdoc3/jsdoc/issues/1132). Moreover, the code causes the latest version of the compiler to error out. 
```js
java.lang.RuntimeException: INTERNAL COMPILER ERROR.
Please report this problem.

CAST node should always have JSDocInfo
	at com.google.common.base.Preconditions.checkNotNull(Preconditions.java:895)
	at com.google.javascript.jscomp.TypeInference.traverse(TypeInference.java:671)
	at com.google.javascript.jscomp.TypeInference.traverseName(TypeInference.java:1269)
	at com.google.javascript.jscomp.TypeInference.traverseDeclarationChild(TypeInference.java:1192)
	at com.google.javascript.jscomp.TypeInference.traverseDeclaration(TypeInference.java:1184)
	at com.google.javascript.jscomp.TypeInference.traverse(TypeInference.java:656)
	at com.google.javascript.jscomp.TypeInference.flowThrough(TypeInference.java:344)
	at com.google.javascript.jscomp.TypeInference.branchedFlowThrough(TypeInference.java:356)
	at com.google.javascript.jscomp.TypeInference.branchedFlowThrough(TypeInference.java:76)
	at com.google.javascript.jscomp.DataFlowAnalysis$BranchedForwardDataFlowAnalysis.flow(DataFlowAnalysis.java:438)
	at com.google.javascript.jscomp.DataFlowAnalysis.analyze(DataFlowAnalysis.java:205)
	at com.google.javascript.jscomp.DataFlowAnalysis.analyze(DataFlowAnalysis.java:173)
	at com.google.javascript.jscomp.TypeInferencePass.inferScope(TypeInferencePass.java:121)
	at com.google.javascript.jscomp.TypeInferencePass$SecondScopeBuildingCallback.enterScope(TypeInferencePass.java:147)
	at com.google.javascript.jscomp.NodeTraversal.pushScope(NodeTraversal.java:1058)
	at com.google.javascript.jscomp.NodeTraversal.pushScope(NodeTraversal.java:1046)
	at com.google.javascript.jscomp.NodeTraversal.traverseWithScope(NodeTraversal.java:443)
	at com.google.javascript.jscomp.TypeInferencePass.inferAllScopes(TypeInferencePass.java:108)
	at com.google.javascript.jscomp.TypeInferencePass.process(TypeInferencePass.java:72)
	at com.google.javascript.jscomp.DefaultPassConfig$47$1.process(DefaultPassConfig.java:1906)
	at com.google.javascript.jscomp.PhaseOptimizer$NamedPass.process(PhaseOptimizer.java:328)
	at com.google.javascript.jscomp.PhaseOptimizer.process(PhaseOptimizer.java:237)
	at com.google.javascript.jscomp.Compiler.check(Compiler.java:1010)
	at com.google.javascript.jscomp.Compiler.performChecksAndTranspilation(Compiler.java:818)
	at com.google.javascript.jscomp.Compiler.lambda$stage1Passes$0(Compiler.java:748)
	at com.google.javascript.jscomp.CompilerExecutor$2.call(CompilerExecutor.java:102)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.NullPointerException: CAST node should always have JSDocInfo
```

kristoferbaxter commented on extensions/amp-story/1.0/amp-story-embedded-component.js:
```diff
@@ -606,7 +614,7 @@ export class AmpStoryEmbeddedComponent {
   updateTooltipBehavior_(target) {
     if (matches(target, LAUNCHABLE_COMPONENTS['a'].selector)) {
-      addAttributesToElement(devAssert(this.tooltip_),
+      addAttributesToElement(dev().assertElement(devAssert(this.tooltip_)),
```
This change seems to be everywhere in this PR. Is there a reason that `devAssert` doesn't just do what it's replaced with here?

rsimha replied:
I don't believe `devAssert` guarantees a non-null `Element`. Without this change:
```js
extensions/amp-story/1.0/amp-story-embedded-component.js:617: ERROR - actual parameter 1 of addAttributesToElement$$module$src$dom does not match formal parameter
found   : (Element|null)
required: Element
      addAttributesToElement(devAssert(this.tooltip_),
                             ^^^^^^^^^^^^^^^^^^^^^^^^
```

kristoferbaxter commented on src/polyfills.js:
```diff
@@ -14,6 +14,8 @@
  * limitations under the License.
  */
 
+/** @fileoverview @suppress {suspiciousCode} */
```
suspiciousCode indeed.

rsimha replied:
@jridgewell wrote the `if(false && ...)` hack. Perhaps it can be removed?

kristoferbaxter commented on src/service/video-manager-impl.js:
```diff
@@ -318,7 +318,7 @@ export class VideoManager {
    *
    * @param {!../video-interface.VideoInterface} video
-   * @return {!../video-interface.VideoInterface} PlayingStates
+   * @return {string}
```
`PlayingStates` sounds far more descriptive than `string`, @alanorozco does this look correct?

rsimha replied:
I'm glad you're keeping me honest here. The return type as originally written seemed meaningless. Probably a copy-pasta typo. Without this change:
```js
src/service/video-manager-impl.js:324: ERROR - inconsistent return type
found   : string
required: VideoInterface$$module$src$video_interface
    return this.getEntryForVideo_(video).getPlayingState();
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

src/service/video-manager-impl.js:1207: ERROR - inconsistent return type
found   : VideoInterface$$module$src$video_interface
required: string
    return this.manager_.getPlayingState(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Unfortunately, using `PlayingStates` (or the fully qualified `!../video-interface.PlayingStates`) as a return type does not work, perhaps because it is a mapping of `string` to `string`, while this function returns just a `string`. Errors seen:

```js
src/service/video-manager-impl.js:321: ERROR - Bad type annotation. Unknown type PlayingStates$$module$src$video_interface
   * @return {!../video-interface.PlayingStates}
```
or
```js
src/service/video-manager-impl.js:321: ERROR - Bad type annotation. Unknown type PlayingStates$$module$src$video_interface
   * @return {PlayingStates}
```

alanorozco replied:
The `VideoInterface` type should probably be fixed instead. I would prefer if we don't lose this type information here.

rsimha replied:
@alanorozco Do you have an alternative in mind? (See comment above for my failed attempts at using `PlayingStates` as a return type.)

rsimha replied:
That worked. Done in https://github.com/ampproject/amphtml/commit/919c7fbd4143768cb1b14e18ed46c2158ff0a44b.

rsimha commented:
@kristoferbaxter Thanks for keeping me honest with your comments. I've left responses to each of them, and will let you resolve them when you're satisfied that there's no better alternative.

Now on to dealing with the single-pass errors...

aghassemi commented on extensions/amp-delight-player/0.1/test/test-amp-delight-player.js:
```diff
@@ -42,7 +42,8 @@ describes.realWin('amp-delight-player', {
 
-  it('renders', () => {
+  // TODO(@aghassemi, #21618): Times out.
+  it.skip('renders', () => {
```
maybe it is a legit failure? if this is green on master and timeout in this PR, there is something going on.

This PR has lots of changes and potential for regression, I strongly feel no tests should be skipped in this PR.

aghassemi commented on extensions/amp-form/0.1/amp-form-textarea.js:
```diff
@@ -113,11 +113,12 @@ export class AmpFormTextarea {
-      }
-    }, MIN_EVENT_INTERVAL_MS);
+    const throttledResize = throttle(
+        /** @type {!Window} */ (this.win_), e => {
```
type should probably be in declaration of the field in constructor instead of usage time

aghassemi commented on extensions/amp-form/0.1/amp-form.js:
```diff
@@ -1068,7 +1068,8 @@ export class AmpForm {
               rendered.setAttribute('i-amphtml-rendered', '');
-              return this.resources_.mutateElement(devAssert(container),
+              return this.resources_.mutateElement(
+                  dev().assertElement(devAssert(container)),
```
nested devAssert could go away

aghassemi commented on extensions/amp-lightbox-gallery/0.1/amp-lightbox-gallery.js:
```diff
@@ -291,7 +291,7 @@ export class AmpLightboxGallery extends AMP.BaseElement {
       return viewer.whenFirstVisible();
     }).then(() => {
-      this.container_ = htmlFor(this.doc_)`
+      this.container_ = htmlFor(/** @type {!Document} */ (this.doc_))`
```
type for `this.doc_ ` in constructor should be updated instead

jridgewell commented on extensions/amp-delight-player/0.1/amp-delight-player.js:
```diff
@@ -205,7 +205,7 @@ class AmpDelightPlayer extends AMP.BaseElement {
     placeholder.firstElementChild.setAttribute('src', src);
 
-    this.placeholderEl_ = placeholder;
+    this.placeholderEl_ = /** @type {HTMLElement} */ (placeholder);
```
Move to the variable init.

rsimha replied:
Done.

jridgewell commented on extensions/amp-form/0.1/amp-form-textarea.js:
```diff
@@ -113,11 +113,12 @@ export class AmpFormTextarea {
-      }
-    }, MIN_EVENT_INTERVAL_MS);
+    const throttledResize = throttle(
+        /** @type {!Window} */ (this.win_), e => {
```
`devAssert`

rsimha replied:
Must be non-null.

aghassemi commented on extensions/amp-lightbox-gallery/0.1/amp-lightbox-gallery.js:
```diff
@@ -374,10 +374,10 @@ export class AmpLightboxGallery extends AMP.BaseElement {
       const metadata = {
-        descriptionText: descText,
-        tagName: clonedNode.tagName,
+        descriptionText: /** @type {string} */ (descText),
```
return type of `getDescription` should handle this. casting doesn't seem right here.

jridgewell commented on extensions/amp-form/0.1/amp-form-textarea.js:
```diff
@@ -230,7 +231,8 @@ export function maybeResizeTextarea(element) {
   return resources.measureMutateElement(element, () => {
-    const computed = computedStyle(win, element);
+    const computed = computedStyle(
+        /** @type {!Window} */ (win), element);
```
Strange.

rsimha replied:
Must be non-null. Without this change:
```js
extensions/amp-form/0.1/amp-form-textarea.js:235: ERROR - actual parameter 1 of computedStyle$$module$src$style does not match formal parameter
found   : (Window|null)
required: Window
        win, element);
        ^^^
```

aghassemi commented on extensions/amp-lightbox-gallery/0.1/amp-lightbox-gallery.js:
```diff
@@ -826,15 +826,15 @@ export class AmpLightboxGallery extends AMP.BaseElement {
 
     return this.mutateElement(() => {
-      setStyles(devAssert(this.carousel_), {
+      setStyles(dev().assertElement(devAssert(this.carousel_)), {
```
nested assert can be removed here and below

jridgewell commented on extensions/amp-iframe/0.1/amp-iframe.js:
```diff
@@ -494,7 +494,8 @@ export class AmpIframe extends AMP.BaseElement {
     if (src !== undefined) {
-      this.iframeSrc = this.transformSrc_(src);
+      this.iframeSrc =
+        /** @type {?string} */ (this.transformSrc_(src));
```
Isn't this already the correct type?

rsimha replied:
`transformSrc_` can return `undefined`. Without this change:
```js
extensions/amp-iframe/0.1/amp-iframe.js:497: ERROR - assignment to property iframeSrc of AmpIframe
found   : (string|undefined)
required: (null|string)
      this.iframeSrc =
      ^^^^^^^^^^^^^^^^
```

aghassemi commented on extensions/amp-lightbox-gallery/0.1/amp-lightbox-gallery.js:
```diff
@@ -925,7 +925,8 @@ export class AmpLightboxGallery extends AMP.BaseElement {
     // a scroll happened.
     this.preventCarouselScrollUnlistener_ = listen(
-        devAssert(this.carousel_), 'scroll', event => {
+        dev().assertElement(devAssert(this.carousel_)),
```
ditto

aghassemi commented on extensions/amp-lightbox/0.1/amp-lightbox.js:
```diff
@@ -497,18 +499,19 @@ class AmpLightbox extends AMP.BaseElement {
-        this.waitForScroll_(this.pos_);
-      }
-    }, 100);
+    this.scrollTimerId_ = /** @type {number} */ (
```
ditto comment about `delay`

aghassemi commented on extensions/amp-live-list/0.1/amp-live-list.js:
```diff
@@ -146,7 +146,7 @@ export class AmpLiveList extends AMP.BaseElement {
     this.isReverseOrder_ = false;
 
-    /** @private @const {!Object<string, string>} */
+    /** @private @const {!Object<string, number>} */
```
technically this should be `time` instead of `number`

aghassemi commented on extensions/amp-playbuzz/0.1/amp-playbuzz.js:
```diff
@@ -197,7 +197,8 @@ class AmpPlaybuzz extends AMP.BaseElement {
       this.iframeLoaded_ = true;
-      this.attemptChangeHeight(this.itemHeight_).catch(() => {/* die */ });
+      this.attemptChangeHeight(
+          /** @type {number} */ (this.itemHeight_)).catch(() => {/* die */ });
```
`assertNumber`

aghassemi commented on extensions/amp-selector/0.1/amp-selector.js:
```diff
@@ -537,7 +537,7 @@ export class AmpSelector extends AMP.BaseElement {
     const {key} = event;
     if (key == Keys.SPACE || key == Keys.ENTER) {
-      if (this.elements_.includes(event.target)) {
+      if (this.elements_.includes(dev().assertElement(event.target))) {
```
I am not convinced `event.target` can't be null here in legit cases.  `includes` don't accept `null`-able`? seems

aghassemi commented on extensions/amp-social-share/0.1/amp-social-share.js:
```diff
@@ -90,7 +90,8 @@ class AmpSocialShare extends AMP.BaseElement {
         getDataParamsFromAttributes(element));
 
-    const hrefWithVars = addParamsToUrl(this.shareEndpoint_, this.params_);
+    const hrefWithVars = addParamsToUrl(
```
`assertSring`

aghassemi commented on extensions/amp-story-auto-ads/0.1/amp-story-auto-ads.js:
```diff
@@ -439,7 +439,7 @@ export class AmpStoryAutoAds extends AMP.BaseElement {
     };
 
-    const configAttrs = this.config_['ad-attributes'];
+    const configAttrs = {'ad-attributes': this.config_['ad-attributes']};
```
this completely changes the code semantics and breaking the loop below

aghassemi commented on extensions/amp-story-auto-ads/0.1/test/test-amp-story-auto-ads.js:
```diff
@@ -241,8 +242,9 @@ describes.realWin('amp-story-auto-ads', {
-    it('should not allow blacklisted attributes', () => {
+  // TODO(@calebcordry, #21618): Fails because ad type is undefined.
+  describe.skip('creation of attributes', () => {
+    it.skip('should not allow blacklisted attributes', () => {
```
this is breaking because of https://github.com/ampproject/amphtml/pull/21618/files#diff-4a080b08c414c51e235cba638bff05baR442 is wrong.

rsimha replied:
Fixed source code, and unskipped both tests in this file.

rsimha commented on extensions/amp-date-display/0.1/amp-date-display.js:
```diff
@@ -281,7 +281,7 @@ export class AmpDateDisplay extends AMP.BaseElement {
   rendered_(element) {
     this.mutateElement(() => {
-      removeChildren(devAssert(this.container_));
+      removeChildren(dev().assertElement(devAssert(this.container_)));
```
Fixed.

rsimha commented on extensions/amp-date-display/0.1/amp-date-display.js:
```diff
@@ -281,7 +281,7 @@ export class AmpDateDisplay extends AMP.BaseElement {
   rendered_(element) {
     this.mutateElement(() => {
-      removeChildren(devAssert(this.container_));
+      removeChildren(dev().assertElement(devAssert(this.container_)));
```
Done.

jridgewell commented on extensions/amp-byside-content/0.1/amp-byside-content.js:
```diff
@@ -173,7 +173,7 @@ export class AmpBysideContent extends AMP.BaseElement {
   layoutCallback() {
     const iframe = this.element.ownerDocument.createElement('iframe');
-    this.iframe_ = iframe;
+    this.iframe_ = /** @type {HTMLIFrameElement} */ (iframe);
```
Is it the `|null` that's throwing it off? Because `HTMLIFrameElement` is an `Element`...

rsimha replied:
It's not just the `null`. The conversion is also a problem.
```js
extensions/amp-byside-content/0.1/amp-byside-content.js:176: ERROR - assignment to property iframe_ of AmpBysideContent
found   : Element
required: (HTMLIFrameElement|null)
    this.iframe_ = iframe;
    ^^^^^^^^^^^^^^^^^^^^^
```

erwinmombay commented on extensions/amp-ad-network-doubleclick-impl/0.1/amp-ad-network-doubleclick-impl.js:
```diff
@@ -530,9 +530,11 @@ export class AmpAdNetworkDoubleclickImpl extends AmpA4A {
-          (this.jsonTargeting && this.jsonTargeting['targeting']) || null,
-          (this.jsonTargeting &&
-            this.jsonTargeting['categoryExclusions']) || null),
+          /** @type {?Object<string, (!Array<string>|string)>} */ (
```
also should this be JsonObject?

rsimha replied:
That works. Done.

erwinmombay commented on extensions/amp-addthis/0.1/amp-addthis.js:
```diff
@@ -286,7 +286,7 @@ class AmpAddThis extends AMP.BaseElement {
     this.applyFillContent(iframe);
     this.element.appendChild(iframe);
-    this.iframe_ = iframe;
+    this.iframe_ = /** @type {HTMLIFrameElement} */ (iframe);
```
is this nullable still? can we mark it as `!`?

rsimha replied:
`this.iframe_` is instantiated with `null`, so yes, it is nullable.

erwinmombay replied:
can we make the denotation explicit by either adding `?` or `!HTMLIframeElement|null`

erwinmombay commented on extensions/amp-form/0.1/amp-form-textarea.js:
```diff
@@ -77,7 +77,7 @@ export class AmpFormTextarea {
 
     /** @private @const */
-    this.win_ = devAssert(this.doc_.defaultView);
+    this.win_ = /** @type {!Window} */ (devAssert(this.doc_.defaultView));
```
can this be on the declaration instead?

rsimha replied:
The problem is that `devAssert` can return `null`. Without this change:
```js
extensions/amp-form/0.1/amp-form-textarea.js:80: ERROR - assignment to property win_ of AmpFormTextarea
found   : (Window|null)
required: Window
    this.win_ = devAssert(this.doc_.defaultView);
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

rsimha commented:
@aghassemi @jridgewell Thanks for the reviews. I've addressed all your comments, and resolved the ones that were either fixed or explained. Feel free to unresolve if there's anything left to do.

@aghassemi I've unskipped all but one of the tests that I initially skipped. (Context: https://twitter.com/cramforce/status/949054860166553600). The only test left to unskip needs a fix based on [this discussion](https://github.com/ampproject/amphtml/pull/21618/files#r270471785).

PTAL and let me know if there's anything left unaddressed.

rsimha commented on extensions/amp-ad-network-doubleclick-impl/0.1/amp-ad-network-doubleclick-impl.js:
```diff
@@ -676,8 +678,9 @@ export class AmpAdNetworkDoubleclickImpl extends AmpA4A {
               !!this.jsonTargeting['targeting'] ?
-                deepMerge(this.jsonTargeting['targeting'],
-                    rewrittenResponse) :
+                deepMerge(/** @type {!Object} */ (
```
Done.

rsimha commented on extensions/amp-ad-network-doubleclick-impl/0.1/sra-utils.js:
```diff
@@ -182,8 +182,10 @@ export function getTargetingAndExclusions(impls) {
-          impl.jsonTargeting['categoryExclusions'] || null));
+          /** @type {?Object<string, (!Array<string>|string)>} */ (
+            impl.jsonTargeting['targeting'] || null),
+          /** @type {?(!Array<string>|string)} */ (
```
Done.

newmuis commented on extensions/amp-story/1.0/amp-story.js:
```diff
@@ -2234,7 +2234,7 @@ export class AmpStory extends AMP.BaseElement {
       getHistoryState(this.win, HistoryState.NAVIGATION_PATH);
     if (historyNavigationPath) {
-      this.storyNavigationPath_ = historyNavigationPath;
+      this.storyNavigationPath_.push(historyNavigationPath);
```
@rsimha this is not logically equivalent, and I think could break history in stories.  A better bet might be `concat`, but before we go there, what was wrong with what was there before?

rsimha replied:
@newmuis Without this change:
```js
extensions/amp-story/1.0/amp-story.js:2237: ERROR - assignment to property storyNavigationPath_ of AmpStory
found   : string
required: (Array<string>|null)
      this.storyNavigationPath_ = historyNavigationPath;
      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

rsimha replied:
I've switched from `push` to `concat` in https://github.com/ampproject/amphtml/pull/21618/commits/66aadedf6c417aa2d25f55b5305edb006b9a760b. Let me know if this works.

rsimha replied:
@newmuis Is this version okay?

rsimha replied:
@newmuis I revisited this code. It's clear that `this.storyNavigationPath_` must be an `{Array<string>}`. This function is `initializeStoryNavigationPath_`, so I'd imagine `this.storyNavigationPath_` is `[]` at this point, so we'd need to add `historyNavigationPath` to an empty array.

rsimha replied:
As discussed in person, I've undone the code changes and fixed the type info. Done in 98a88e7.

erwinmombay commented on extensions/amp-analytics/0.1/transport.js:
```diff
@@ -140,7 +140,7 @@ export class Transport {
    * @param {!Window} win
    * @param {!Element} element
-   * @param {!../../../src/preconnect.Preconnect|undefined} opt_preconnect
+   * @param {(!../../../src/preconnect.Preconnect|undefined)=} opt_preconnect
```
i think you can remove `undefined` if you add `=`

rsimha replied:
Good catch. Done.

erwinmombay commented on extensions/amp-apester-media/0.1/amp-apester-media.js:
```diff
@@ -94,7 +94,7 @@ class AmpApesterMedia extends AMP.BaseElement {
      */
     this.mediaId_ = null;
-    /** @private {Array<Function>} */
+    /** @private {!Array<Function>} */
```
`{!Array<function()>}` would be more appropriate

rsimha replied:
Done.

erwinmombay commented on extensions/amp-autocomplete/0.1/amp-autocomplete.js:
```diff
@@ -145,7 +145,7 @@ export class AmpAutocomplete extends AMP.BaseElement {
     userAssert(inputElements.length === 1,
         `${TAG} should contain exactly one <input> child`);
-    this.inputElement_ = inputElements[0];
+    this.inputElement_ = /** @type {HTMLInputElement} */ (inputElements[0]);
```
still nullable?

rsimha replied:
Made it non-nullable.

erwinmombay commented on extensions/amp-brightcove/0.1/amp-brightcove.js:
```diff
@@ -443,17 +444,17 @@ class AmpBrightcove extends AMP.BaseElement {
   /** @override */
   getPlayedRanges() {
-    return this.playedRanges_;
+    return /** @type {!Array<Array<number>>} */ (this.playedRanges_);
```
is the inner Array nullable?

rsimha replied:
Made it non-nullable.

jridgewell commented on extensions/amp-accordion/0.1/amp-accordion.js:
```diff
@@ -548,7 +548,8 @@ class AmpAccordion extends AMP.BaseElement {
       }
-      const newFocusHeader = this.headers_[newFocusIndex];
+      const newFocusHeader =
+      dev().assertElement(this.headers_[newFocusIndex]);
```
These `dev().assertX` calls are blockers because of the code size increases. We need to find a way to fix `devAssert` to get these type assertions and allow it to be eliminated from the build.

rsimha replied:
The problem is that `devAssert` merely verifies the truishness of the argument, and knows nothing about its type.

Without this change:
```js
extensions/amp-accordion/0.1/amp-accordion.js:552: ERROR - actual parameter 1 of tryFocus$$module$src$dom does not match formal parameter
found   : Node
required: Element
      tryFocus(newFocusHeader);
               ^^^^^^^^^^^^^^
```

With `devAssert` instead of `dev().assertElement` (same error):
```js
extensions/amp-accordion/0.1/amp-accordion.js:552: ERROR - actual parameter 1 of tryFocus$$module$src$dom does not match formal parameter
found   : Node
required: Element
      tryFocus(newFocusHeader);
               ^^^^^^^^^^^^^^
```

jridgewell replied:
`headers_` should be changed to point to `Element` then?

jridgewell commented on extensions/amp-analytics/0.1/visibility-manager-for-mapp.js:
```diff
@@ -107,7 +107,8 @@ export class VisibilityManagerForMApp extends VisibilityManager {
     // By the time `#getRootLayoutBox` is called, it is guaranteed that
     // onVisibilityChangeHandler has been called at least once
-    return devAssert(this.intersectionRect_);
+    return /** @type {!../../../src/layout-rect.LayoutRectDef} */ (
```
`devAssert` has been broken by this update then, which we need to fix. It should be able to remove the falsey values from the type assertion without us having to cast.

rsimha replied:
I tried changing the `@return` type of `devAssert` from `{T}` to `{!T}` but that didn't help. Do you have any ideas for how to guarantee that `devAssert` will return a non-null value? This should probably go in as a separate PR.

rsimha replied:
I tried changing the `@return` type of `devAssert` from `{T}` to `{!T}` but that didn't help. Do you have any ideas for how to guarantee that `devAssert` will return a non-null value? This should probably go in as a separate PR.

rsimha commented on extensions/amp-ad-network-doubleclick-impl/0.1/amp-ad-network-doubleclick-impl.js:
```diff
@@ -530,9 +530,11 @@ export class AmpAdNetworkDoubleclickImpl extends AmpA4A {
-            this.jsonTargeting['categoryExclusions']) || null),
+          /** @type {?Object<string, (!Array<string>|string)>} */ (
+            (this.jsonTargeting && this.jsonTargeting['targeting']) || null),
+          /** @type {?(!Array<string>|string)} */ (
```
Done.

rsimha commented:
Thanks everyone for your patience. This PR is now ready for a final review.

Here are the things I've manually verified:

- `gulp check-types` is green
- `gulp dist --fortesting` (multi-pass, with all extensions) is green
- Manually tested multi-pass build with a few `examples/` pages
- `gulp test --integration --nobuild --compiled --headless` (integration tests against multi-pass build) is green
- `gulp dist --fortesting --single_pass --pseudo_names` (single-pass, with all extensions) is green
- Manually testsed single-pass build with a few `examples/` pages
- `gulp test --integration --nobuild --compiled --single_pass --headless` (integration tests against single-pass build) is green
- Travis unminified build is green ([link](https://travis-ci.org/ampproject/amphtml/jobs/515320641))
- Travis checks are green ([link](https://travis-ci.org/ampproject/amphtml/jobs/515320642))
- Travis single-pass build and tests are green ([link](https://travis-ci.org/ampproject/amphtml/jobs/515320644))
- Travis unit and integration tests are green ([link](https://travis-ci.org/ampproject/amphtml/jobs/515320646))

rsimha commented on extensions/amp-ad-network-doubleclick-impl/0.1/amp-ad-network-doubleclick-impl.js:
```diff
@@ -191,11 +191,11 @@ export class AmpAdNetworkDoubleclickImpl extends AmpA4A {
 
-    /** @type {number} */
-    this.adKey = 0;
+    /** @type {string} */
```
@aghassemi @jridgewell Is this change from `number` to `string` okay?

aghassemi replied:
looks fine to me. can't find anything that relies on this being number

rsimha commented on extensions/amp-analytics/0.1/visibility-manager-for-mapp.js:
```diff
@@ -107,7 +107,8 @@ export class VisibilityManagerForMApp extends VisibilityManager {
     // onVisibilityChangeHandler has been called at least once
-    return devAssert(this.intersectionRect_);
+    return /** @type {!../../../src/layout-rect.LayoutRectDef} */ (
+      devAssert(this.intersectionRect_));
```
@jridgewell Let me know if you can think of a fix for `devAssert` returning a `null|{T}`. Until then, we're going to have to typecast its return value, or use alternatives like `dev().assertElement()`.

rsimha replied:
@kristoferbaxter The size increases pointed out by you and @jridgewell are coming from changes of this sort. The problem is that `devAssert` returns the type `null|{T}`, which breaks when a non-null value is expected. Here, I was able to fix it with a cast. In other places, I had to use `dev().assertElement` because casting to `Element` is forbidden by presubmit checks.

Let me know if you can think of a fix for `devAssert`, in which case I can revert most of the changes that cause a size increase.

rsimha commented on extensions/amp-mustache/0.1/amp-mustache.js:
```diff
@@ -33,9 +34,9 @@ const TAG = 'amp-mustache';
+ * @extends {BaseTemplate}
  */
-export class AmpMustache extends AMP.BaseTemplate {
+export class AmpMustache extends BaseTemplate {
```
@erwinmombay Is this okay? `AMP.BaseTemplate` is invisible without this change.

jridgewell replied:
We cannot do this, it will cause a large increase in file size.

rsimha replied:
This was fixed in https://github.com/ampproject/amphtml/pull/21618/commits/23f71344953a48da94bd9f734a7156daac74589a.

kristoferbaxter commented:
2.89KB increase seems quite odd. Are there new patterns increasing binary size?

aghassemi commented on extensions/amp-story/0.1/test/test-amp-story-hint.js:
```diff
@@ -45,7 +45,8 @@ describes.fakeWin('amp-story hint layer', {}, env => {
 
-  it('should be able to show navigation help overlay', () => {
+  // TODO(@gmajoulet, #21618): Fails in AmpStoryHint.showHint_.
+  it.skip('should be able to show navigation help overlay', () => {
```
are we sure this is not a regression? I am strongly against skipping any tests in this PR.

rsimha replied:
This PR doesn't touch the `amp-story-hint` component, and these tests pass when run in isolation. However, when run with other tests, there's this error:

```
     ● should be able to show navigation help overlay
        HeadlessChrome 73.0.3683 (Linux 0.0.0)
      TypeError: hostElement.attachShadow is not a function
          at createShadowRoot (usr/local/google/home/rsimha/src/amphtml/src/shadow-embed.js:73:30)
          at createShadowRootWithStyle (usr/local/google/home/rsimha/src/amphtml/extensions/amp-story/0.1/utils.js:99:22)
          at AmpStoryHint.build (usr/local/google/home/rsimha/src/amphtml/extensions/amp-story/0.1/amp-story-hint.js:165:5)
          at AmpStoryHint.showHint_ (usr/local/google/home/rsimha/src/amphtml/extensions/amp-story/0.1/amp-story-hint.js:190:10)
          at AmpStoryHint.showNavigationOverlay (usr/local/google/home/rsimha/src/amphtml/extensions/amp-story/0.1/amp-story-hint.js:215:10)
          at Context.<anonymous> (usr/local/google/home/rsimha/src/amphtml/extensions/amp-story/0.1/test/test-amp-story-hint.js:52:18)
```
I reached out to @gmajoulet to see if he could spot a relation between the changes in this PR and the test failures. Note that since they are unit tests, they are run on components without running closure compiler, and do not use minified output.

As of now, these are the only tests that fail with this PR, and @gmajoulet said he'd look into this. Let me know if you spot something I haven't thought of.

aghassemi commented:
@rsimha Can we hold on merging until the release for AMPConf is cut? I am worried about a big change like this making it to the release that's schedule for the day before AMPConf.

rsimha commented on extensions/amp-apester-media/0.1/test/test-amp-apester-media.js:
```diff
@@ -150,7 +150,8 @@ describes.realWin(
       });
-      it('render playlist', () => {
+      // TODO(@aghassemi, #21618): Fails because mediaAttribute_ isn't a string.
+      it.skip('render playlist', () => {
```
Done in https://github.com/ampproject/amphtml/commit/a0d8a000bb441d47af92effd8a069f16e3675d5f. Unskipped test.

rsimha commented on extensions/amp-animation/0.1/amp-animation.js:
```diff
@@ -95,9 +95,11 @@ export class AmpAnimation extends AMP.BaseElement {
-    this.configJson_ = tryParseJson(scriptElement.textContent, error => {
-      throw user().createError('failed to parse animation script', error);
-    });
+    this.configJson_ = /** @type {?JsonObject} */ (
```
Fixed in https://github.com/ampproject/amphtml/pull/21618/commits/7cc19f024bf8e6e396b694144d0a6f8f1e5133b3.

rsimha commented on extensions/amp-animation/0.1/amp-animation.js:
```diff
@@ -95,9 +95,11 @@ export class AmpAnimation extends AMP.BaseElement {
-      throw user().createError('failed to parse animation script', error);
-    });
+    this.configJson_ = /** @type {?JsonObject} */ (
+      tryParseJson(scriptElement.textContent, error => {
```
Changed `tryParseJson` to return `null` in the failing case instead of `undefined`. With this, there's no need for the cast in this file. Done in https://github.com/ampproject/amphtml/pull/21618/commits/7cc19f024bf8e6e396b694144d0a6f8f1e5133b3.

jridgewell commented on src/gesture.js:
```diff
@@ -409,7 +409,7 @@ export class Gestures {
     const overserver = this.overservers_[recognizer.getType()];
     if (overserver) {
-      overserver.fire(new Gesture(recognizer.getType(), data, Date.now(),
+      overserver.fire(new Gesture(recognizer.getType(), data, new Date(),
```
This is a different API, and has perf implications. `Date` is much more expensive to create than the timestamp. If no one is actually using this, let's just delete it instead.

jridgewell commented on extensions/amp-story/1.0/amp-story.js:
```diff
@@ -2269,7 +2269,8 @@ export class AmpStory extends AMP.BaseElement {
     if (historyNavigationPath) {
-      this.storyNavigationPath_ = historyNavigationPath;
+      this.storyNavigationPath_ =
+          this.storyNavigationPath_.concat([historyNavigationPath]);
```
This is a change in behavior

newmuis replied:
Actually, can we leave this code the way it was initially and just change the type of `getHistoryState` to `{*}` (which is what it actually should be)?

rsimha replied:
Tried this, but I'm seeing new errors elsewhere in the code:
```js
extensions/amp-story/1.0/amp-story.js:885: ERROR - inconsistent return type
found   : *
required: string
      return historyPage;
             ^^^^^^^^^^^

extensions/amp-story/1.0/amp-story.js:2272: ERROR - assignment to property storyNavigationPath_ of AmpStory
found   : *
required: (Array<string>|null)
      this.storyNavigationPath_ = historyNavigationPath;
      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

rsimha replied:
As discussed in person, I've undone the code changes and fixed the type info. Done in https://github.com/ampproject/amphtml/pull/21618/commits/98a88e70b4c2b6c229b1d987d37a8aca65eeda51.

jridgewell commented on extensions/amp-mustache/0.2/amp-mustache.js:
```diff
@@ -29,9 +30,9 @@ const TAG = 'amp-mustache';
+ * @extends {BaseTemplate}
  */
-export class AmpMustache extends AMP.BaseTemplate {
+export class AmpMustache extends BaseTemplate {
```
This'll be a large increase in file size!

aghassemi commented:
sweet, thanks @rsimha

jridgewell commented:
I got `devAssert()` and `dev().assertElement` removals working:

```diff
diff --git a/build-system/runner/src/org/ampproject/AmpCommandLineRunner.java b/build-system/runner/src/org/ampproject/AmpCommandLineRunner.java
index 53cf7f088..b6cf93632 100644
--- a/build-system/runner/src/org/ampproject/AmpCommandLineRunner.java
+++ b/build-system/runner/src/org/ampproject/AmpCommandLineRunner.java
@@ -50,11 +50,17 @@ public class AmpCommandLineRunner extends CommandLineRunner {
   /**
    * List of string suffixes to eliminate from the AST.
    */
-  ImmutableMap<String, Set<String>> suffixTypes = ImmutableMap.of(
-      "module$src$log.dev", ImmutableSet.of(
-          "assert", "fine", "assertElement", "assertString",
-          "assertNumber", "assertBoolean", "assertArray"),
-      "module$src$log.user", ImmutableSet.of("fine"));
+  ImmutableSet<String> suffixTypes = ImmutableSet.of(
+      "dev$$module$src$log().assert()",
+      "dev$$module$src$log().fine()",
+      "dev$$module$src$log().assertElement()",
+      "dev$$module$src$log().assertString()",
+      "dev$$module$src$log().assertNumber()",
+      "dev$$module$src$log().assertArray()",
+      "dev$$module$src$log().assertBoolean()",
+      "devAssert$$module$src$log()",
+      "user$$module$src$log().fine()"
+      );
 
 
   ImmutableMap<String, Node> assignmentReplacements = ImmutableMap.of(
diff --git a/build-system/runner/src/org/ampproject/AmpPass.java b/build-system/runner/src/org/ampproject/AmpPass.java
index 5ce98d284..feb59510a 100644
--- a/build-system/runner/src/org/ampproject/AmpPass.java
+++ b/build-system/runner/src/org/ampproject/AmpPass.java
@@ -15,10 +15,12 @@
  */
 package org.ampproject;
 
+import java.util.LinkedList;
 import java.util.Map;
 import java.util.Set;
 
 import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.ImmutableMap;
 import com.google.javascript.jscomp.AbstractCompiler;
 import com.google.javascript.jscomp.HotSwapCompilerPass;
 import com.google.javascript.jscomp.NodeTraversal;
@@ -43,14 +45,14 @@ import com.google.javascript.rhino.Node;
 class AmpPass extends AbstractPostOrderCallback implements HotSwapCompilerPass {
 
   final AbstractCompiler compiler;
-  private final Map<String, Set<String>> stripTypeSuffixes;
-  private final Map<String, Node> assignmentReplacements;
-  private final Map<String, Node> prodAssignmentReplacements;
+  private final ImmutableSet<String> stripTypeSuffixes;
+  private final ImmutableMap<String, Node> assignmentReplacements;
+  private final ImmutableMap<String, Node> prodAssignmentReplacements;
   final boolean isProd;
 
   public AmpPass(AbstractCompiler compiler, boolean isProd,
-        Map<String, Set<String>> stripTypeSuffixes,
-        Map<String, Node> assignmentReplacements, Map<String, Node> prodAssignmentReplacements) {
+        ImmutableSet<String> stripTypeSuffixes,
+        ImmutableMap<String, Node> assignmentReplacements, ImmutableMap<String, Node> prodAssignmentReplacements) {
     this.compiler = compiler;
     this.stripTypeSuffixes = stripTypeSuffixes;
     this.isProd = isProd;
@@ -166,59 +168,15 @@ class AmpPass extends AbstractPostOrderCallback implements HotSwapCompilerPass {
   }
 
   /**
-   * For a function that looks like:
-   * function fun(val) {
-   *   return dev().assert(val);
-   * }
-   *
-   * The AST would look like:
-   * RETURN 24 [length: 25] [source_file: ./src/main.js]
-   *   CALL 24 [length: 17] [source_file: ./src/main.js]
-   *     GETPROP 24 [length: 12] [source_file: ./src/main.js]
-   *       CALL 24 [length: 5] [source_file: ./src/main.js]
-   *         NAME $dev$$module$src$log$$ 38 [length: 3] [originalname: dev] [source_file: ./src/log.js]
-   *         STRING assert 24 [length: 6] [source_file: ./src/main.js]
-   *     NAME $val$$ 24 [length: 3] [source_file: ./src/main.js]
-   *
-   * We are looking for the `CALL` that has a child NAME "$dev$$module$src$log$$" (or any signature from keys)
-   * and a child STRING "assert" (or any other signature from Set<String> value)
    */
   private boolean isCallRemovable(Node n) {
     if (n == null || !n.isCall()) {
       return false;
     }
 
-    Node callGetprop = n.getFirstChild();
-    if (callGetprop == null || !callGetprop.isGetProp()) {
-      return false;
-    }
-
-    // Temp hack not needed in single pass compiler because it can inline the
-    // this itself.
-    if ("module$src$log.devAssert".equals(callGetprop.getQualifiedName())) {
-      return true;
-    }
-
-    Node parentCall = callGetprop.getFirstChild();
-    if (parentCall == null || !parentCall.isCall()) {
-      return false;
-    }
-
-    Node parentCallGetprop = parentCall.getFirstChild();
-    Node methodName = parentCall.getNext();
-    if (parentCallGetprop == null || !parentCallGetprop.isGetProp() ||
-        methodName == null || !methodName.isString()) {
-      return false;
-    }
-
-    String parentMethodName = parentCallGetprop.getQualifiedName();
-    Set<String> methodCallNames = stripTypeSuffixes.get(parentMethodName);
-    if (methodCallNames == null) {
-      return false;
-    }
-
-    for (String methodCallName : methodCallNames) {
-      if (methodCallName == methodName.getString()) {
+    String name = buildQualifiedName(n);
+    for (String removable : stripTypeSuffixes) {
+      if (name.equals(removable)) {
         return true;
       }
     }
@@ -278,6 +236,35 @@ class AmpPass extends AbstractPostOrderCallback implements HotSwapCompilerPass {
     return false;
   }
 
+  /**
+   * Builds a string representation of MemberExpression and CallExpressions.
+   */
+  private String buildQualifiedName(Node n) {
+    StringBuilder sb = new StringBuilder();
+    buildQualifiedNameInternal(n, sb);
+    return sb.toString();
+  }
+
+  private void buildQualifiedNameInternal(Node n, StringBuilder sb) {
+    if (n == null) {
+      sb.append("NULL");
+      return;
+    }
+
+    if (n.isCall()) {
+      buildQualifiedNameInternal(n.getFirstChild(), sb);
+      sb.append("()");
+    } else if (n.isGetProp()) {
+      buildQualifiedNameInternal(n.getFirstChild(), sb);
+      sb.append(".");
+      buildQualifiedNameInternal(n.getSecondChild(), sb);
+    } else if (n.isName() || n.isString()) {
+      sb.append(n.getString());
+    } else {
+      sb.append("UNKNOWN");
+    }
+  }
+
   private void replaceWithBooleanExpression(boolean bool, Node n, Node parent) {
     Node booleanNode = bool ? IR.trueNode() : IR.falseNode();
     booleanNode.useSourceInfoIfMissingFrom(n);
diff --git a/build-system/runner/test/org/ampproject/AmpPassTest.java b/build-system/runner/test/org/ampproject/AmpPassTest.java
index 04fc79baa..1499290c5 100644
--- a/build-system/runner/test/org/ampproject/AmpPassTest.java
+++ b/build-system/runner/test/org/ampproject/AmpPassTest.java
@@ -19,10 +19,11 @@ import org.junit.Test;
  */
 public class AmpPassTest extends CompilerTestCase {
 
-  ImmutableMap<String, Set<String>> suffixTypes = ImmutableMap.of(
-      "module$src$log.dev",
-          ImmutableSet.of("assert", "fine", "assertElement", "assertString", "assertNumber"),
-      "module$src$log.user", ImmutableSet.of("fine"));
+  ImmutableSet<String> suffixTypes = ImmutableSet.of(
+      "dev$$module$src$log().assert()",
+      "dev$$module$src$log().fine()",
+      "devAssert$$module$src$log()"
+      );
 
   ImmutableMap<String, Node> assignmentReplacements = ImmutableMap.of(
       "IS_MINIFIED",
@@ -46,26 +47,46 @@ public class AmpPassTest extends CompilerTestCase {
     test(
         LINE_JOINER.join(
              "(function() {",
-             "  var module$src$log = { dev: function() { return { fine: function() {}} } };",
-             "  module$src$log.dev().fine('hello world');",
+             "  dev$$module$src$log().fine('hello world');",
              "  console.log('this is preserved');",
             "})()"),
         LINE_JOINER.join(
              "(function() {",
-             "  var module$src$log = { dev: function() { return { fine: function() {}} } };",
              "  'hello world';",
              "  console.log('this is preserved');",
             "})()"));
     test(
         LINE_JOINER.join(
              "(function() {",
-             "  var module$src$log = { dev: function() { return { fine: function() {}} } };",
-             "  module$src$log.dev().fine();",
+             "  dev$$module$src$log().fine();",
+             "  console.log('this is preserved');",
+            "})()"),
+        LINE_JOINER.join(
+             "(function() {",
+             "  console.log('this is preserved');",
+            "})()"));
+  }
+
+  @Test public void testDevAssertRemoval() throws Exception {
+    test(
+        LINE_JOINER.join(
+             "(function() {",
+             "  devAssert$$module$src$log('hello world');",
+             "  console.log('this is preserved');",
+            "})()"),
+        LINE_JOINER.join(
+             "(function() {",
+             "  'hello world';",
+             "  console.log('this is preserved');",
+            "})()"));
+    test(
+        LINE_JOINER.join(
+             "(function() {",
+             "  devAssert$$module$src$log();",
              "  console.log('this is preserved');",
             "})()"),
         LINE_JOINER.join(
              "(function() {",
-             "  var module$src$log = { dev: function() { return { fine: function() {}} } };",
              "  console.log('this is preserved');",
             "})()"));
   }
@@ -90,26 +111,22 @@ public class AmpPassTest extends CompilerTestCase {
     test(
         LINE_JOINER.join(
              "(function() {",
-             "  var module$src$log = { dev: function() { return { assert: function() {}} } };",
-             "  module$src$log.dev().assert('hello world');",
+             "  dev$$module$src$log().assert('hello world');",
              "  console.log('this is preserved');",
             "})()"),
         LINE_JOINER.join(
              "(function() {",
-             "  var module$src$log = { dev: function() { return { assert: function() {}} } };",
              "  \"hello world\";",
              "  console.log('this is preserved');",
             "})()"));
     test(
         LINE_JOINER.join(
              "(function() {",
-             "  var module$src$log = { dev: function() { return { assert: function() {}} } };",
-             "  var someValue = module$src$log.dev().assert();",
+             "  var someValue = dev$$module$src$log().assert();",
              "  console.log('this is preserved', someValue);",
             "})()"),
         LINE_JOINER.join(
              "(function() {",
-             "  var module$src$log = { dev: function() { return { assert: function() {}} } };",
              "  var someValue;",
              "  console.log('this is preserved', someValue);",
             "})()"));
@@ -119,13 +136,11 @@ public class AmpPassTest extends CompilerTestCase {
     test(
         LINE_JOINER.join(
              "(function() {",
-             "  var module$src$log = { dev: function() { return { assert: function() {}} } };",
-             "  var someValue = module$src$log.dev().assert(true, 'This is an error');",
+             "  var someValue = dev$$module$src$log().assert(true, 'This is an error');",
              "  console.log('this is preserved', someValue);",
             "})()"),
         LINE_JOINER.join(
              "(function() {",
-             "  var module$src$log = { dev: function() { return { assert: function() {}} } };",
              "  var someValue = true;",
              "  console.log('this is preserved', someValue);",
             "})()"));
@@ -134,14 +149,12 @@ public class AmpPassTest extends CompilerTestCase {
         LINE_JOINER.join(
              "(function() {",
              "  function add(a, b) { return a + b; }",
-             "  var module$src$log = { dev: function() { return { assert: function() {}} } };",
-             "  var someValue = add(module$src$log.dev().assert(3), module$src$log.dev().assert(3));",
+             "  var someValue = add(dev$$module$src$log().assert(3), dev$$module$src$log().assert(3));",
              "  console.log('this is preserved', someValue);",
             "})()"),
         LINE_JOINER.join(
              "(function() {",
              "  function add(a, b) { return a + b; }",
-             "  var module$src$log = { dev: function() { return { assert: function() {}} } };",
              "  var someValue = add(3, 3);",
              "  console.log('this is preserved', someValue);",
             "})()"));
@@ -152,14 +165,12 @@ public class AmpPassTest extends CompilerTestCase {
         // Does reliasing
         LINE_JOINER.join(
              "(function() {",
-             "  var module$src$log = { dev: function() { return { assert: function() {}} } };",
-             "  var someValue = module$src$log.dev().assert;",
+             "  var someValue = dev$$module$src$log().assert;",
              "  console.log('this is preserved', someValue);",
             "})()"),
         LINE_JOINER.join(
              "(function() {",
-             "  var module$src$log = { dev: function() { return { assert: function() {}} } };",
-             "  var someValue = module$src$log.dev().assert;",
+             "  var someValue = dev$$module$src$log().assert;",
              "  console.log('this is preserved', someValue);",
             "})()"));
   }
diff --git a/build-system/runner/test/org/ampproject/AmpPassTestEnvTest.java b/build-system/runner/test/org/ampproject/AmpPassTestEnvTest.java
index 63776150f..3550c06ce 100644
--- a/build-system/runner/test/org/ampproject/AmpPassTestEnvTest.java
+++ b/build-system/runner/test/org/ampproject/AmpPassTestEnvTest.java
@@ -4,6 +4,7 @@ package org.ampproject;
 import java.util.Set;
 
 import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
 import com.google.javascript.rhino.IR;
 import com.google.javascript.rhino.Node;
 import com.google.javascript.jscomp.Compiler;
@@ -17,7 +18,7 @@ import org.junit.Test;
  */
 public class AmpPassTestEnvTest extends CompilerTestCase {
 
-  ImmutableMap<String, Set<String>> suffixTypes = ImmutableMap.of();
+  ImmutableSet<String> suffixTypes = ImmutableSet.of();
   ImmutableMap<String, Node> assignmentReplacements = ImmutableMap.of(
       "IS_MINIFIED",
       IR.trueNode());
```

erwinmombay commented:
@jridgewell kudos, cleaned up that code so much. 👍

rsimha commented:
@jridgewell Bravo! I've included your patch in https://github.com/ampproject/amphtml/pull/21618/commits/eca2401250543f58254fc02dd89b80603167a632. With it, the bundlesize of `v0.js` goes down from 88.53KB to 84.97KB. We're better off than we were prior to this upgrade.

erwinmombay commented on extensions/amp-story-auto-ads/0.1/amp-story-auto-ads.js:
```diff
@@ -188,10 +188,10 @@ export class AmpStoryAutoAds extends AMP.BaseElement {
     this.ampStory_ = null;
 
-    /** @private {?../../amp-story/0.1/navigation-state.NavigationState} */
+    /** @private {?../../amp-story/1.0/navigation-state.NavigationState} */
```
is the 1.0 path change intentional?

erwinmombay commented on src/custom-element.js:
```diff
@@ -103,8 +103,12 @@ function isTemplateTagSupported() {
 export function createCustomElementClass(win, name) {
-  const baseCustomElement = createBaseCustomElementClass(win);
-  /** @extends {HTMLElement} */
+  const baseCustomElement = /** @type {Function} */ (
```
Function is usually not allowed, this should probably be something like `function(new:WhateverTheTypeIs)`

rsimha replied:
Here's the error I was trying to fix:
```js
src/custom-element.js:111: ERROR - Bad type annotation. Unknown type baseCustomElement
  class CustomAmpElement extends baseCustomElement {
                                 ^
```
I used `{Function}` because it's used as a return type for `createCustomElementClass` and `createBaseCustomElementClass`. If `Function` needs to be changed to `function(new:WhateverTheTypeIs)`, it probably needs to be done in a lot more places. Okay if we punt that to a separate PR? (I want to avoid roping in too many other changes in this already huge PR.)

https://github.com/ampproject/amphtml/blob/4a9ec7291763d35789b6e31a40b6d3b2c13ccb80/src/custom-element.js#L98-L105

https://github.com/ampproject/amphtml/blob/4a9ec7291763d35789b6e31a40b6d3b2c13ccb80/src/custom-element.js#L129-L135

jridgewell commented on extensions/amp-mustache/0.2/amp-mustache.js:
```diff
@@ -24,14 +24,20 @@ import mustache from '../../../third_party/mustache/mustache';
+ * @suppress {checkTypes}
  */
-export class AmpMustache extends AMP.BaseTemplate {
+export class AmpMustache extends BaseTemplateDef {
```
This will break the inheritance chain. It must be `AMP.BaseTemplate` at runtime.

rsimha replied:
Gotcha. I'm going to revert all changes to this file. The only problem that remains is to satisfy the type checker. Using `@suppress {checkTypes, reportUnknownTypes}` did not work. Still looking for an alternative solution.

```js
extensions/amp-mustache/0.1/amp-mustache.js:39: ERROR - Bad type annotation. Unknown type AMP.BaseTemplate
export class AmpMustache extends AMP.BaseTemplate {
                                 ^

extensions/amp-mustache/0.2/amp-mustache.js:35: ERROR - Bad type annotation. Unknown type AMP.BaseTemplate
export class AmpMustache extends AMP.BaseTemplate {
                                 ^
```

rsimha replied:
@erwinmombay Any ideas for how to make `AMP.BaseTemplate` visible to `AmpMustache` without importing `BaseTemplate`?

jridgewell replied:
Can't you typecast `/** @type {BaseTemplate$$module$src$service$template_impl} */(AMP.BaseTemplate)`. This should work the same as extending from `AMP.BaseElement`, which is also not defined in extensions.

rsimha replied:
Yep, typedef'ing `AMP.BaseTemplate` worked. Done in https://github.com/ampproject/amphtml/pull/21618/commits/9268ec8e2011790c6d134bf1804978dd86a0acdd.

rsimha commented:
@aghassemi @newmuis Let me know if you have any more comments. I plan on merging this after next week's canary cut.

aghassemi commented on extensions/amp-form/0.1/amp-form-textarea.js:
```diff
@@ -77,7 +77,7 @@ export class AmpFormTextarea {
 
     /** @private @const */
-    this.win_ = devAssert(this.doc_.defaultView);
+    this.win_ = /** @type {!Window} */ (devAssert(this.doc_.defaultView));
```
all the dev asserts in this file are pretty pointless and can be removed.

rsimha replied:
Unfortunately, they're needed. Without them:
```js
extensions/amp-form/0.1/amp-form-textarea.js:116: ERROR - actual parameter 1 of throttle$$module$src$utils$rate_limit does not match formal parameter
found   : (Window|null)
required: Window
    const throttledResize = throttle(this.win_, e => {
                                     ^^^^^^^^^

extensions/amp-form/0.1/amp-form-textarea.js:233: ERROR - actual parameter 1 of computedStyle$$module$src$style does not match formal parameter
found   : (Window|null)
required: Window
    const computed = computedStyle(win, element);
                                   ^^^

extensions/amp-form/0.1/amp-form-textarea.js:290: ERROR - actual parameter 1 of Resources$$module$src$service$resources_impl.prototype.measureMutateElement does not match formal parameter
found   : (HTMLBodyElement|null)
required: Element
  return resources.measureMutateElement(body, () => {
                                        ^^^^

extensions/amp-form/0.1/amp-form-textarea.js:291: ERROR - actual parameter 1 of computedStyle$$module$src$style does not match formal parameter
found   : (Window|null)
required: Window
    const computed = computedStyle(win, textarea);
                                   ^^^

extensions/amp-form/0.1/amp-form-textarea.js:305: ERROR - actual parameter 1 of Resources$$module$src$service$resources_impl.prototype.measureMutateElement does not match formal parameter
found   : (HTMLBodyElement|null)
required: Element
    return resources.measureMutateElement(body, () => {
                                          ^^^^
```

aghassemi commented on extensions/amp-jwplayer/0.1/amp-jwplayer.js:
```diff
@@ -82,12 +82,10 @@ class AmpJWPlayer extends AMP.BaseElement {
         '';
-    this.contentContextual_ = element.getAttribute('data-content-contextual') ||
-        false;
+    this.contentContextual_ = element.hasAttribute('data-content-contextual');
```
this would behave differently for `data-content-contextual=false` (since the value gets passed as query string param) which per documentation is valid. I suggest changing types to string and using `''` as default.

rsimha replied:
Done.

rsimha commented:
The canary cut for the AMP Conf release has been released: https://github.com/ampproject/amphtml/releases/tag/1904091426070 (cc @choumx who is release on-duty this week)

Once @aghassemi and @newmuis approve the final changes, I'll do another full round of testing and then merge this PR.

rsimha commented on extensions/amp-brightcove/0.1/amp-brightcove.js:
```diff
@@ -118,11 +118,12 @@ class AmpBrightcove extends AMP.BaseElement {
-          'Did not receive ready callback from player %s.' +
-        ' Ensure it has the videojs-amp-support plugin.', this.playerId_);
-    }, 3000);
+    this.readyTimeout_ = /** @type {number} */ (
```
I've logged #21804 to track this.

rsimha commented:
This PR is now ready to merge. For posterity, here is a list of things I've reviewed / verified / tested.

**Code review:**
- Self-reviewed all 240+ files in this PR to make sure there are no stray changes
- Re-read all 200+ review comments to make sure all of them were satisfactorily addressed and resolved
- Filed issues for future clean up tasks / feature requests

**Local tests:**
- Ran `gulp check-types` to make sure there are no type errors
- Ran `gulp dist --fortesting` to make sure minified multi-pass build is healthy
    - Ran `gulp serve --compiled` and manually tested several components in multi-pass mode
- Ran `gulp dist --fortesting --single_pass --pseudo_names` to make sure minified single-pass build is healthy
    - Ran `gulp serve --compiled` and manually tested several components in single-pass mode
- Ran `gulp test --local-changes --headless` to make sure all unit tests affected by this PR are green

**Travis tests:**
- Checks (`presubmit`, `lint`, `dep-check`, and `check-types`) are green ([link](https://travis-ci.org/ampproject/amphtml/jobs/518418541#L640-L745))
- Un-minified build is green ([link](https://travis-ci.org/ampproject/amphtml/jobs/518418540#L640-L644))
- Visual tests are green with zero diffs due to this PR ([link](https://percy.io/ampproject/amphtml/builds/1719711))
- All unit tests are green ([link](https://travis-ci.org/ampproject/amphtml/jobs/518418545#L3772))
- All integration tests are green ([link](https://travis-ci.org/ampproject/amphtml/jobs/518418545#L3742))

**AMP Release schedule**
- AMP Conference 2019 is next week. We're doing a [release](https://github.com/ampproject/amphtml/issues/21788) on Tuesday, Apr 16 using this week's canary cut, which was based on the [1904091426070](https://github.com/ampproject/amphtml/releases/tag/1904091426070) tag.
- This PR was merged after the AMP Conf release tag, so the main AMP Conf release is unaffected by any of these changes.

rsimha commented:
![image](https://user-images.githubusercontent.com/26553114/55906634-b15cf680-5ba2-11e9-8092-4e2e6ae2f7ab.png)

dvoytenko commented:
@rsimha Probably needs to be rolled back :( It updated the third_party code improperly. E.g. `third_party/config/config.js` cannot do `dev().assert`.

rsimha commented:
@dvoytenko Thanks for your comment. We can revert the changes to `third_party/`, as you've done in #21827. For an explanation of why I (incorrectly) made the change, see https://github.com/ampproject/amphtml/pull/21827#pullrequestreview-225734306.

I'll approve and merge your partial revert ASAP.
